### PR TITLE
Use Spotify recommendations API in getRelatedTracks()

### PR DESCRIPTION
**Pull request by mariusbegby**
## Changes
- Updates `getRelatedTracks()` method of `SpotifyExtractor` to use new `internal.getRelatedTracks()` method.
- Implemented Spotify `internal.getRelatedTracks()` method, which fetches related tracks based on [Spotify recommendations API](https://developer.spotify.com/documentation/web-api/reference/get-recommendations).
- Updated internal `SpotifyTrack` interface `artists` property to add new property `uri` which is on extractor interface `SpotifySong`.

## Notes

The Spotify recommendations API takes tracks, artists and genres for seeding (data basis) when providing recommendations in response. Due to the limitations of the extractor `getRelatedTracks()` method only getting a single track as input, the API call is only seeded with one track and the related artist. Therefore, with the new behavior for `getRelatedTracks()` in `SpotifyExtractor` introduced by this change, it is more likely that the autoplay will "drift off" and start playing less relevant tracks as time goes on.

To solve this: If we can extend the `SpotifyExtractor` method `getRelatedTracks()` to take in multiple tracks as input, this seeding can be improved, reducing the chance that Spotify recommendations API provides less relevant results in response.

## Status

- [x] These changes have been tested and formatted properly.
- [ ] This PR includes only documentation changes, no code change.
- [ ] This PR introduces some Breaking changes.

### CODE REVIEW

**Comment by mariusbegby** [comment]
I think we should try and find a solution to the issues mentioned under notes before merging this.

The implementation of using the Spotify recommendations API for `getRelatedTracks()` works, but it doesn't necessarily work better than previous behavior due to lack of seeding.

**Comment by retrouser955** [comment]
You could cache the last 5 tracks from the `stream` function and use it as seed for the `getRelatedTracks` function.

**Comment by twlite** [comment]
> You could cache the last 5 tracks from the `stream` function and use it as seed for the `getRelatedTracks` function.

I think it makes more sense to let extractors access queue history object instead of having to implement yet another cache.

File: packages/extractor/src/extractors/SpotifyExtractor.ts
Comment (twlite) [inline packages/extractor/src/extractors/SpotifyExtractor.ts:53]: why is this value hardcoded?

File: packages/extractor/src/extractors/SpotifyExtractor.ts
Comment (twlite) [inline packages/extractor/src/extractors/SpotifyExtractor.ts:68]: We dont need to fetch 5 tracks here, autoplay only needs 1 track.

File: packages/extractor/src/internal/Spotify.ts
Comment (twlite) [inline packages/extractor/src/internal/Spotify.ts:283]: so how do we seed multiple tracks here? joining by `TRACK_ID,TRACK_ID`? What if rest of the tracks in history are not from spotify?

File: packages/extractor/src/internal/Spotify.ts
Comment (twlite) [inline packages/extractor/src/internal/Spotify.ts:322]: would not this break existing code that depend on this type? I think it is better to revert this? 🤔

**Review by twlite — CHANGES_REQUESTED** [review]

File: packages/extractor/src/extractors/SpotifyExtractor.ts
Comment (twlite) [reply packages/extractor/src/extractors/SpotifyExtractor.ts:53]: I think it will be better to first attempt getting related tracks using spotify api and if response is null, we fall back to default search method.

Comment (mariusbegby) [reply packages/extractor/src/extractors/SpotifyExtractor.ts:53]: I agree we can fall back to old behaviour if response is null.

The value is hardcoded just to easier compare new/old behaviour during development, it will of course be removed.

File: packages/extractor/src/extractors/SpotifyExtractor.ts
Comment (mariusbegby) [reply packages/extractor/src/extractors/SpotifyExtractor.ts:68]: Yeah. Just used this to easily see quality of recommendations from Spotify API. Final implementation will have 1 as limit hardcoded. This parameter will be removed.

File: packages/extractor/src/internal/Spotify.ts
Comment (mariusbegby) [reply packages/extractor/src/internal/Spotify.ts:283]: Yes, correct. It is a comma separated list of track ids.

We will have to first verify that they are Spotify track IDs before getting used in the API call.

The recommendations API call only allow Spotify IDs, and needs at least one.

File: packages/extractor/src/internal/Spotify.ts
Comment (mariusbegby) [reply packages/extractor/src/internal/Spotify.ts:322]: Although the interface is exported, I could not see it used anywhere yet, only in internal Spotify.ts

The id field for an artist is not even used in internal Spotify.ts as far as I could see, only the name property.

The actual response from Spotify API looks like this:
```
[
  {
    external_urls: {
      spotify: 'https://open.spotify.com/artist/5Matrg5du62bXwer29cU5T'
    },
    href: 'https://api.spotify.com/v1/artists/5Matrg5du62bXwer29cU5T',
    id: '5Matrg5du62bXwer29cU5T',
    name: 'Benny The Butcher',
    type: 'artist',
    uri: 'spotify:artist:5Matrg5du62bXwer29cU5T'
  }
]
```

Since this SpotifyTrack is like a DTO for Spotify track response, I think it's correct to use uri.

Without it, I will have to create a new interface for SpotifyTrack to return this:
```
artists: m.artists.map((m) => {
    return {
        name: m.name,
        uri: m.uri
    };
}),
```
Because m.uri is infact correct, that's the response from Spotify API.

Comment (mariusbegby) [reply packages/extractor/src/internal/Spotify.ts:322]: Although the interface is exported, I could not see it used anywhere yet, only in internal Spotify.ts

The id field for an artist is not even used in internal Spotify.ts as far as I could see, only the name property.

The actual response from Spotify API looks like this:
```
[
  {
    external_urls: {
      spotify: 'https://open.spotify.com/artist/5Matrg5du62bXwer29cU5T'
    },
    href: 'https://api.spotify.com/v1/artists/5Matrg5du62bXwer29cU5T',
    id: '5Matrg5du62bXwer29cU5T',
    name: 'Benny The Butcher',
    type: 'artist',
    uri: 'spotify:artist:5Matrg5du62bXwer29cU5T'
  }
]
```

Why I did not use the id property is because the SpotifyExtractor SpotifySong interface only has these properties:
```
artists: {
    name: string;
    uri: string;
}[];
```

I will update the internal SpotifyTrack interface to add the uri property instead of replacing it:
```
artists: {
    id: string;
    name: string;
    uri: string;
}[];
```

**Comment by mariusbegby** [comment]
> > You could cache the last 5 tracks from the `stream` function and use it as seed for the `getRelatedTracks` function.
> 
> I think it makes more sense to let extractors access queue history object instead of having to implement yet another cache.

I agree with this approach. 

For a first iteration, should all extractors be updated to receive the queue history, then they could only use the last track in history for now. And for SpotifyExtractor we can use last 5 Spotify tracks from history?

File: packages/extractor/src/internal/Spotify.ts
Comment (mariusbegby) [reply packages/extractor/src/internal/Spotify.ts]: We have to decide on a way to accept multiple tracks in the Extractor getRelatedTracks() method. Then we can update the internal API accordingly to use multiple tracks.

**Comment by twlite** [comment]
> For a first iteration, should all extractors be updated to receive the queue history, then they could only use the last track in history for now. And for SpotifyExtractor we can use last 5 Spotify tracks from history?

I will push an update that adds history support for extractors in a while.

**Comment by twlite** [comment]
@mariusbegby I have pushed new changes, you can git pull and implement the update. To get basic idea on how it works, take a look at this

https://github.com/Androz2091/discord-player/blob/e417681998cede5e2a3f7ff551659488abd5865b/packages/extractor/src/extractors/SoundCloudExtractor.ts#L52

File: packages/extractor/src/internal/Spotify.ts
Comment (mariusbegby) [reply packages/extractor/src/internal/Spotify.ts]: Now takes multiple track ids and artist ids. Spotify recommendations API has a combined max limit of 5, so I also added logic to handle that.

**Comment by retrouser955** [comment]
> > You could cache the last 5 tracks from the `stream` function and use it as seed for the `getRelatedTracks` function.
> 
> I think it makes more sense to let extractors access queue history object instead of having to implement yet another cache.

I strongly agree with this implementation.

**Comment by mariusbegby** [comment]
@twlite and me tested out this new implementation of Spotify recommendations API.

While it technically works, Spotify gives pretty bad recommendations that easily go to unrelated tracks.

Need to put this on hold and see if it can be tuned to give better, more relevant tracks as recommendations.

**Comment by mariusbegby** [comment]
Closing this, but still keeping this in mind for possible future improvements. Unfortunately the Spotify Recommendations API doesn't work too well for this use case.

**Pull request closed without merging**
